extensions/attributegen/plugin.cc

```diff
@@ -0,0 +1,147 @@
+/* Copyright 2020 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/attributegen/plugin.h"
+
+// WASM_PROLOG
+#ifndef NULL_PLUGIN
+
+#else  // NULL_PLUGIN
+
+#include "extensions/common/wasm/null/null.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace AttributeGen {
+
+bool PluginRootContext::onConfigure(size_t) {
+  std::unique_ptr<WasmData> configuration = getConfiguration();
+  // Parse configuration JSON string.
+  JsonParseOptions json_options;
+  json_options.ignore_unknown_fields = true;
+  istio::attributegen::PluginConfig config;
+  Status status =
+      JsonStringToMessage(configuration->toString(), &config, json_options);
+  if (status != Status::OK) {
+    LOG_WARN(absl::StrCat(
+        "Cannot parse plugin configuration JSON string [YAML not supported]: ",
+        configuration->toString()));
+    return false;
+  }
+
+  return initAttributeGen(config);
+}
+
+bool PluginRootContext::initAttributeGen(
+    const istio::attributegen::PluginConfig& config) {
+  // Clear prior host resources.
+  // TODO(mjog) see if this still needed since onDone is always called.
+  cleanupAttributeGen();
+
+  for (auto agconfig : config.attributes()) {
+    EvalPhase phase = on_log;
+    if (agconfig.phase() == istio::attributegen::ON_REQUEST) {
+      phase = on_request;
+    }
+    auto ag = AttributeGenerator(phase, agconfig.output_attribute());
+    for (auto matchconfig : agconfig.match()) {
+      uint32_t token = 0;
+      if (matchconfig.condition().empty()) {
+        ag.add_match(Match("", 0, matchconfig.value()));
+        continue;
+      }
+      if (createExpression(matchconfig.condition(), &token) != WasmResult::Ok) {
+        LOG_WARN(absl::StrCat("Cannot create expression: <",
+                              matchconfig.condition(), "> for ",
+                              agconfig.output_attribute()));
+        return false;
```

Comment:
why fail fast and not ignore?

Reply:
I do not follow, what do you want to ignore?

Reply:
Generally we never want to reject config unless something terrible happens since it immediately crashes envoy in some cases. Seems like one bad expression is not a sufficient reason to crash.